File: Microsoft.Toolkit.Uwp.Input.GazeInteraction/ComboBoxItemGazeTargetItem.cpp
#include "pch.h"
#include "ComboBoxItemGazeTargetItem.h"
#include "ComboBoxItemGazeTargetItem.g.cpp"
#include "winrt/Microsoft.UI.Xaml.Automation.Peers.h"
#include "winrt/Microsoft.UI.Xaml.Automation.Provider.h"

using namespace winrt;
using namespace winrt::Microsoft::UI::Xaml::Automation::Peers;
using namespace winrt::Microsoft::UI::Xaml::Automation::Provider;

namespace winrt::Microsoft::Toolkit::Uwp::Input::GazeInteraction::implementation
{
    void ComboBoxItemGazeTargetItem::Invoke()
    {
        auto peer = FrameworkElementAutomationPeer::FromElement(TargetElement());
        ComboBoxItemAutomationPeer comboBoxItemAutomationPeer{ peer.try_as<ComboBoxItemAutomationPeer>() };
        ComboBoxItem comboBoxItem{ comboBoxItemAutomationPeer.Owner().try_as<ComboBoxItem>() };

        AutomationPeer ancestor = comboBoxItemAutomationPeer;
        ComboBoxAutomationPeer comboBoxAutomationPeer{ ancestor.try_as<ComboBoxAutomationPeer>() };
        while (comboBoxAutomationPeer == nullptr)
        {
            ancestor = ancestor.Navigate(AutomationNavigationDirection::Parent).try_as<AutomationPeer>();
            comboBoxAutomationPeer = ancestor.try_as<ComboBoxAutomationPeer>();
        }

        comboBoxItem.IsSelected(true);
        comboBoxAutomationPeer.Collapse();
    }
}

File: Microsoft.Toolkit.Uwp.Input.GazeInteraction/ComboBoxItemGazeTargetItem.cs
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

using Windows.UI.Xaml;
using Windows.UI.Xaml.Automation.Peers;
using Windows.UI.Xaml.Controls;

namespace Microsoft.Toolkit.Uwp.Input.GazeInteraction
{
    internal class ComboBoxItemGazeTargetItem : GazeTargetItem
    {
        internal ComboBoxItemGazeTargetItem(UIElement element)
            : base(element)
        {
        }

        internal override void Invoke()
        {
            var peer = FrameworkElementAutomationPeer.FromElement(TargetElement);
            var comboBoxItemAutomationPeer = peer as ComboBoxItemAutomationPeer;
            var comboBoxItem = (ComboBoxItem)comboBoxItemAutomationPeer.Owner;

            AutomationPeer ancestor = comboBoxItemAutomationPeer;
            var comboBoxAutomationPeer = ancestor as ComboBoxAutomationPeer;
            while (comboBoxAutomationPeer == null)
            {
                ancestor = ancestor.Navigate(AutomationNavigationDirection.Parent) as AutomationPeer;
                comboBoxAutomationPeer = ancestor as ComboBoxAutomationPeer;
            }

            comboBoxItem.IsSelected = true;
            comboBoxAutomationPeer.Collapse();
        }
    }
}